lib/matplotlib/axes/_axes.py

```diff
@@ -7632,38 +7639,15 @@ def hist(self, x, bins=None, range=None, density=False, weights=None,
         labels = [] if label is None else np.atleast_1d(np.asarray(label, str))
 
         if histtype == "step":
-            ec = kwargs.get('edgecolor', colors)
-        else:
-            ec = kwargs.get('edgecolor', None)
-        if ec is None or cbook._str_lower_equal(ec, 'none'):
-            edgecolors = itertools.repeat(ec)
-        else:
-            edgecolors = itertools.cycle(mcolors.to_rgba_array(ec))
-
-        fc = kwargs.get('facecolor', colors)
-        if cbook._str_lower_equal(fc, 'none'):
-            facecolors = itertools.repeat(fc)
-        else:
-            facecolors = itertools.cycle(mcolors.to_rgba_array(fc))
+            kwargs.setdefault('edgecolor', colors)
 
-        hatches = itertools.cycle(np.atleast_1d(kwargs.get('hatch', None)))
-        linewidths = itertools.cycle(np.atleast_1d(kwargs.get('linewidth', None)))
-        if 'linestyle' in kwargs:
-            linestyles = itertools.cycle(mlines._get_dash_patterns(kwargs['linestyle']))
-        else:
-            linestyles = itertools.repeat(None)
+        kwargs, style_gen = _style_helpers.style_generator(kwargs)
 
         for patch, lbl in itertools.zip_longest(patches, labels):
             if not patch:
                 continue
             p = patch[0]
-            kwargs.update({
-                'hatch': next(hatches),
-                'linewidth': next(linewidths),
-                'linestyle': next(linestyles),
-                'edgecolor': next(edgecolors),
-                'facecolor': next(facecolors),
-            })
+            kwargs.update(next(style_gen))
```

Comment:
This is the correct translation keeping existing behavior.

Semi-OT: I'm wondering whether the behavior is logically correct: If `patches` is empty, we do not advance the style. This means depending on data contents the assignment of styles to datasets  may vary.

Reply:
`bar()` and `grouped_bar()` require the same length:

https://github.com/matplotlib/matplotlib/blob/4aeb773422464799998d900198b35cb80e94b3e1/lib/matplotlib/axes/_axes.py#L2601-L2603

https://github.com/matplotlib/matplotlib/blob/89aa37129ffe4296c7f93727e897b2f34f1127fe/lib/matplotlib/axes/_axes.py#L3331

We should do the same for stackplot. If people do not want to label selected elements they should mark this explicitly in the list of labels.

Reply:
Having said the above, the "unused labels" test came in at #28073, whereas the `if patch` check goes all the way back to https://github.com/matplotlib/matplotlib/commit/30fddc3d58fa0ac753e6ec4d7e27a3cb8a935889.  So there may be reasons that are not covered by tests :face_with_peeking_eye: